INTERNALS.md

```diff
@@ -23,75 +23,188 @@ Serves endpoints through 3 controllers:
+
+All timestamps are represented as integers, milliseconds since the Unix epoch.
+
 
 `topics`
 
   The set of all topic names.
 
 `subscribers`
 
-  The set of all subscriber UUIDs.
+  The set of all subscriber tokens.
 
-`topics:{uuid}`
+`topics:{token}`
 
-  The set of topic names subscribed to by subscriber `uuid`.
+  The set of topic names subscribed to by subscriber `{token}`.
 
-`subscribers:{topic}`
+`subscribers:{name}`
 
-  The set of subscriber UUIDs having subscribed to topic `name`.
+  The set of subscriber tokens having subscribed to topic `{name}`.
 
 `topic:{name}`
 
   A hash containing metadata has about a topic. Keys:
   - `publisher`: the UUID of the (singly authorized) publisher
   - `counter`: the cumulative number of events received
 
-`subscriber:{uuid}`
+`subscriber:{token}`
 
   A hash of subscription medatata. Keys:
   - `callback`: the URL to send events to.
-  - `timeout`: how long to defer event delivery for batching purposes.
+  - `timeout`: how long to defer event delivery for batching purposes (aka deadline).
   - `max_events`: maximum number of events to batch.
   - `uuid`: the credential to use when delivering events.
 
-`queue:new:{subscriber}`
+`batch:{bid}` (list)
+
+  A list whose first items are:
+  - the subscriber token this batch is for,
+  - the timestamp at which the batch was created,
+  - the number of delivery attempts for this batch.
+  followed by the serialized messages to deliver. 
+  
+  `bid` is the batch's UID.
+
+`batches:current:{token}` (set)
+
+  The current batch(es) for subscriber `{token}`, if any.
+  Values are batch UIDs.
+
+  This should normally have at most 1 element, although under high
+  concurrency situations multiple batches may be created simultaneously.
+  
+`batches:index` (sorted set)
+
+  An index for the set of all batches.
+  The score is the batch's creation timestamp; values are batch UIDs.
+
+`batches:counters:event` (hash)
+
+  Number of currently extant events, per subscriber name.
+
+`batches:counters:batch` (hash)
+
+  Number of currently extant batches, per subscriber name.
+
+`jobs:index:{q}` (set)
+
+  All jobs currently in the queue named `{q}`, in any state. Jobs are
+  represented as a MessagePack-encoded array of 2 elements, the job name and its
+  array of arguments.
+
+`jobs:scheduled:{q}` (sorted set)
+
+  Jobs scheduled for later execution, scored by their deadline timestamp.
+  Jobs a represented as above.
+
+`jobs:queue:{q}` (list)
 
-  A list of UIDs of messages to be delivered, in reception order.
+  Jobs in the "instant" state, for immediate execution.
 
-`queue:pending:{subscriber}`
+`jobs:pending:{q}:{worker}` (list)
 
-  A zset of UIDs of messages for which delivery is in progress, keyed by the
-  timestamp of the attempt.
-  This gets cleared when messages are acked or nacked.
+  A list of at most 1 element, containing the job currently being processed by
+  worker `{worker}`.
```

Comment:
Because that allows us to `BRPOPLPUSH` (block-pop from the queue, push to the worker's "pending list"), which is interrupt-safe (ie. you'll never lose jobs).
This is what, for instance, Sidekiq Pro uses internally.